src/main/resources/dbreplica/migration/V1.0.5__task_assignments_plsql.sql

```diff
@@ -0,0 +1,120 @@
+l_jurisdiction text, l_location text, l_role_category text, l_task_name text)
+ RETURNS bigint
+ LANGUAGE plpgsql
+AS $function$
+
+declare
+l_update_id bigint;
+l_end_reason text;
+l_new_assignment boolean;
+
+begin
+
+    l_end_reason =
+    case
+        when l_update_action in ('UnassignAssign','UnassignClaim','UnclaimAssign','AutoUnassignAssign')
+            then 'REASSIGNED'
+        when l_update_action in ('Unclaim')
+            then 'UNCLAIMED'
+        when l_update_action in ('AutoUnassign','Unassign')
+            then 'UNASSIGNED'
+        when l_update_action in ('AutoCancel','Cancel')
+            then 'CANCELLED'
+        else l_state
+    end;
+
+    if ((COALESCE(l_assignee,'') <> '')
+      and (l_update_action in ('AutoUnassignAssign', 'UnassignAssign', 'UnassignClaim', 'UnclaimAssign', 'Assign', 'AutoAssign', 'Claim'))) then
+
+        select not exists(select 1 from cft_task_db.task_assignments where task_id = l_task_id and assignee = l_assignee
+                                and (assignment_end is null or assignment_start = l_updated))
+        into l_new_assignment;
+
+        if (l_new_assignment) then
+            update cft_task_db.task_assignments
+                set  assignment_end =  l_updated,
+                     assignment_end_reason = l_end_reason
+                where task_id = l_task_id and assignment_end is null and assignment_start < l_updated;
+
+            insert into cft_task_db.task_assignments
+                (assignment_start, assignee, task_id, service, location, role_category, task_name)
+              values
+                (l_updated, l_assignee, l_task_id, l_jurisdiction, l_location, l_role_category, l_task_name)
+                returning assignment_id into l_update_id;
+        end if;
+    end if;
+
+    if ((COALESCE(l_assignee,'') <> '') and (l_update_action in ('Complete', 'AutoCancel', 'Cancel')))
+        or ((COALESCE(l_assignee,'') = '') and (l_update_action in ('Unassign', 'Unclaim', 'AutoUnassign'))) then
+
+        update cft_task_db.task_assignments
+            set  assignment_end =  l_updated,
+                 assignment_end_reason = l_end_reason
+            where task_id = l_task_id and assignment_end is null and assignment_start < l_updated;
+    end if;
+
+return l_update_id;
+end $function$;
+
+--
+-- Function to call from triggers whenever a task record is inserted or updated.
+--
+create or replace function cft_task_db.on_task_history_insert()
+  returns trigger
+  language plpgsql
+as $function$
+begin
+  perform cft_task_db.add_task_assignments(new.task_id);
+return new;
+end $function$;
+
+DROP TRIGGER IF EXISTS trg_on_task_history_insert ON cft_task_db.task_history;
+--
+CREATE TRIGGER trg_on_task_history_insert after insert on cft_task_db.task_history
+                                                           for each row execute function cft_task_db.on_task_history_insert();
```

Comment:
I think this trigger would be better after insert on cft_task_db.tasks.  It is at this point that technically the replication has completed since we are inserting into the task_history _before_ it is inserted into the replica Tasks table.